docs/design/2021-03-30-small-table-caching.md

```diff
@@ -0,0 +1,174 @@
+# Proposal: Caching a table in memory
+
+- Author(@tiancaiamao)
+- Discussion PR: https://github.com/pingcap/tidb/pull/23673
+- Tracking Issue: https://github.com/pingcap/tidb/issues/25293
+
+## Introduction
+
+For a small, frequently visited, and rarely changed table, caching the whole table in memory in the TiDB server can improve performance.
+
+## Motivation or Background
+
+For tables that are too small, they are located in just one region, that region would become a hotspot, and such hotspots would cause a performance bottleneck. By directly caching the small table data in the TiDB layer, such hotspot issues can be solved.
+
+We are caching tables for some scenarios already, in a way we're not realized. For example, our handling of global variables. When the client has a new connection request, TiDB needs to load global variables. After a downtime or restart in some situations, all clients will reconnect the TiDB server simultaneously, loading global variables becomes a hotspot. A special processing of TiDB is to cache global variables, and the modification of global variables will take 3 seconds to take effect. The data of privilege related tables are also cached.
```

Comment:
May I ask where you got the 3 seconds from? There used to be a cache feature which was 2 seconds, but since TiDB ~5.1/5.2 there has been a sysvar cache. It reloads every 30 seconds, but when I profiled it -- it took a lot less time than 2 seconds. If it takes this long, we should investigate it, since it presumbly will hurt thoughput on low core count machines.

Reply:
Actually - I didn't realise the date. The proposal is 1 year old, and this text likely describes the old behavior. The point is still valid in user cases, as well as the table `mysql.tidb` which some system variables much read from for legacy reasons.

Reply:
> The proposal is 1 year old, and this text likely describes the old behavior.

Exactly!

Reply:
Lets make it accurate just so future readers understand current context:
```suggestion
We are caching tables for some scenarios already, in a way we're not realized. For example, our handling of global variables. To prevent row-at-a-time loading of each system variable from TiKV, we needed to implement the [sysvar cache](https://github.com/pingcap/tidb/pull/24359). This helps reduce `show variables like ..` latency, but the data set is small and the problem is very generic. It also doesn't help in cases where system variables read from `mysql.tidb` instead, which does not have a cache. The data of privilege related tables are also cached.
```